### PR TITLE
[ContainerApp] support pull image using env msi

### PR DESCRIPTION
**Pull request by njuCZ**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [ ] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally? (`pip install wheel==0.30.0` required)
- [ ] My extension version conforms to the [Extension version schema](https://github.com/Azure/azure-cli/blob/release/doc/extensions/versioning_guidelines.md)

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
ContainerApp

File: src/containerapp/HISTORY.rst [hunk 1]
```diff
@@ -4,6 +4,10 @@ Release History
 ===============
 upcoming
 ++++++
+* 'az containerapp create': Support --registry-identity use Environment identity for ACR authentication.
```
Comment (Greedygre) [inline src/containerapp/HISTORY.rst:7]:
```suggestion
* 'az containerapp create': Support --registry-identity to use Environment identity for ACR authentication.
```

Same as below

File: src/containerapp/azext_containerapp/_params.py [hunk 2]
```diff
@@ -32,6 +32,7 @@ def load_arguments(self, _):
         c.argument('build_env_vars', nargs='*', help="A list of environment variable(s) for the build. Space-separated values in 'key=value' format.",
                    validator=validate_build_env_vars, is_preview=True)
         c.argument('max_inactive_revisions', type=int, help="Max inactive revisions a Container App can have.", is_preview=True)
+        c.argument('registry_identity', help="The managed identity with which to authenticate to the Azure Container Registry (instead of username/password). Use 'system' for a system-defined identity, Use 'system-environment' for an environment level system-defined identity or a resource id for a user-defined environment/containerapp level identity. The managed identity should have been assigned acrpull permissions on the ACR before deployment (use 'az role assignment create --role acrpull ...').")
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/_params.py:35]: argument `registry_identity` is only used by `az containerapp create` and `az containerapp job create`, this argument can update under to avoild duplicate code:
`with self.argument_context('containerapp') as c:`

**Comment by Greedygre** [comment]
Please ensure CI pass, please rerun related tests:
```
FAILED src/containerapp/azext_containerapp/tests/latest/test_containerapp_commands.py::ContainerappRegistryIdentityTests::test_containerapp_identity_registry
FAILED src/containerapp/azext_containerapp/tests/latest/test_containerapp_scenario.py::ContainerappScenarioTest::test_containerapp_registry_msi
FAILED src/containerapp/azext_containerapp/tests/latest/test_containerappjob_with_identity.py::ContainerAppJobsCRUDOperationsTest::test_containerappjob_identity_registry
```

File: src/containerapp/azext_containerapp/custom.py [hunk 3]
```diff
@@ -3105,3 +3111,111 @@ def create_dotnet_component(cmd, dotnet_component_name, environment_name, resour
...
+                containerapp_def["properties"]["configuration"]["secrets"],
+                username,
+                server,
+                password,
+                update_existing_secret=True)
+        else:
+            registry["identity"] = identity
+
+        registries_def.append(registry)
+
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/custom.py:3187]: Please add some comments about which code you copy from azure-cli and didn't change.

File: src/containerapp/azext_containerapp/containerapp_job_decorator.py [hunk 4]
```diff
@@ -500,8 +514,190 @@ def update(self):
...
+
+        if self.get_argument_registry_identity():
+            if is_registry_msi_system(self.get_argument_registry_identity()):
+                set_managed_identity(self.cmd, self.get_argument_resource_group_name(), self.containerapp_def, system_assigned=True)
+            elif is_valid_resource_id(self.get_argument_registry_identity()):
+                parsed_managed_env = parse_resource_id(self.get_argument_managed_env())
+                managed_env_name = parsed_managed_env['name']
+                managed_env_rg = parsed_managed_env['resource_group']
+                if not env_has_managed_identity(self.cmd, managed_env_rg, managed_env_name, self.get_argument_registry_identity()):
+                    set_managed_identity(self.cmd, self.get_argument_resource_group_name(), self.containerapp_def, user_assigned=[self.get_argument_registry_identity()])
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/containerapp_job_decorator.py:697]: you use self.containerapp_def, which is not correct here.

File: src/containerapp/azext_containerapp/containerapp_job_decorator.py [hunk 5]
```diff
@@ -500,8 +514,190 @@ def update(self):
...
+
+        if self.get_argument_workload_profile_name():
+            self.containerappjob_def["properties"]["workloadProfileName"] = self.get_argument_workload_profile_name()
+            ensure_workload_profile_supported(self.cmd, managed_env_name, managed_env_rg, self.get_argument_workload_profile_name(),
+                                              managed_env_info)
+
+        if self.get_argument_registry_identity():
+            if is_registry_msi_system(self.get_argument_registry_identity()):
+                set_managed_identity(self.cmd, self.get_argument_resource_group_name(), self.containerapp_def, system_assigned=True)
+            elif is_valid_resource_id(self.get_argument_registry_identity()):
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/containerapp_job_decorator.py:692]: `elif is_valid_resource_id(self.get_argument_registry_identity()):`?

If the identity value is `system-environment`, how it run into line 696?

File: src/containerapp/azext_containerapp/containerapp_job_decorator.py [hunk 6]
```diff
@@ -500,8 +514,190 @@ def update(self):
...
+        self.containerappjob_def["properties"]["configuration"] = config_def
+        self.containerappjob_def["properties"]["template"] = template_def
+        self.containerappjob_def["tags"] = self.get_argument_tags()
+
+        if self.get_argument_workload_profile_name():
+            self.containerappjob_def["properties"]["workloadProfileName"] = self.get_argument_workload_profile_name()
+            ensure_workload_profile_supported(self.cmd, managed_env_name, managed_env_rg, self.get_argument_workload_profile_name(),
+                                              managed_env_info)
+
+        if self.get_argument_registry_identity():
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/containerapp_job_decorator.py:689]: Recommend to change the logic you changed as following, 
```
    def set_up_create_acr_pull_role_assignment(self):
        if self.get_argument_registry_identity() and not is_registry_msi_system(self.get_argument_registry_identity()) and not is_registry_msi_system_environment(self.get_argument_registry_identity()):
            logger.info("Creating an acrpull role assignment for the registry identity")
            create_acrpull_role_assignment(self.cmd, self.get_argument_registry_server(), self.get_argument_registry_identity(), skip_error=True)

    def set_up_registry_identity(self):
        if self.get_argument_registry_identity():
            if is_registry_msi_system(self.get_argument_registry_identity()):
```

```
# preview logic
        self.set_up_registry_identity()
```

File: src/containerapp/azext_containerapp/tests/latest/test_containerappjob_with_identity.py [hunk 7]
```diff
@@ -162,4 +163,78 @@ def test_containerapp_eventjob_identity_keda(self, resource_group):
...
+        self.cmd(f'containerapp job show -g {resource_group} -n {job}', checks=[
+            JMESPathCheck("properties.provisioningState", "Succeeded"),
+            JMESPathCheck("identity.type", "None"),
+            JMESPathCheck("properties.configuration.registries[0].server", f"{acr}.azurecr.io"),
+            JMESPathCheck("properties.configuration.registries[0].identity", "system-environment", case_sensitive=False),
+            JMESPathCheck("properties.template.containers[0].image", image_name),
+        ])
+
+        # update use env user assigned identity
+        self.cmd(f'containerapp job registry set -g {resource_group} -n {job} --server {acr}.azurecr.io --identity {user_identity_id}')
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/tests/latest/test_containerappjob_with_identity.py:222]: Can we use `az containerapp job registry set --identity system-environment`? If yes, we should have test cover it.

File: src/containerapp/azext_containerapp/_params.py [hunk 8]
```diff
@@ -374,6 +379,9 @@ def load_arguments(self, _):
         c.argument('name', name_type, id_part=None, help="The name of the Containerapp.")
         c.argument('resource_group_name', arg_type=resource_group_name_type, id_part=None)
 
+    with self.argument_context('containerapp job registry') as c:
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/_params.py:382]: Recommend to add example in `help` for system-environment

File: src/containerapp/azext_containerapp/_help.py [hunk 9]
```diff
@@ -2014,3 +2014,27 @@
...
+"""
+
+helps['containerapp job registry set'] = """
+    type: command
+    short-summary: Add or update a container registry's details in a Container App Job.
+    examples:
+    - name: Configure a Container App Job to use environment system assigned managed identity to authenticate Azure container registry.
+      text: |
+          az containerapp job registry set -n my-containerapp-job -g MyResourceGroup \\
+              --server MyContainerappJobRegistry.azurecr.io --identity system-environment
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/_help.py:2039]: You removed the original:
az containerapp job registry set -n my-containerapp-job -g MyResourceGroup \\
              --server MyContainerappJobRegistry.azurecr.io --username MyRegistryUsername --password MyRegistryPassword

**Review by Greedygre — APPROVED** [review]

**Review by zhoxing-ms — APPROVED** [review]

**Pull request merged**